File: hybridset.js
class Multiset {
    constructor(iterable) {
        if (iterable instanceof Map) {
            this.elements = new Map(
                Array.from(iterable.entries()).filter(
                    ([element, multiplicity]) => multiplicity > 0
                )
            );
        }
        if (iterable instanceof Multiset) {
            this.elements = new Map(iterable.elements);
        }
        if (iterable instanceof Set) {
            this.elements = new Map();
            iterable.forEach((e) => {
                this.elements.set(e, 1);
            });
        }
        if (iterable instanceof Array) {
            this.elements = new Map();
            iterable.forEach((e) => {
                if (!this.elements.has(e)) {
                    this.elements.set(e, 1);
                } else {
                    this.elements.set(e, this.elements.get(e) + 1);
                }
            });
        }
    }
    copy() {
        return new Multiset(this);
    }
    isEmpty() {
        return this.elements.size == 0;
    }
    toString() {
        if (this.isEmpty()) {
            return "{}";
        }
        var s = "{";
        this.elements.forEach((multiplicity, element) => {
            s += `${element}, `.repeat(multiplicity);
        });
        return s.slice(0, -2) + "}";
    }
    clear() {
        this.elements.clear();
    }
    length() {
        // let len = 0;
        return Array.from(this.elements.values()).reduce((x, y) => x + y);
        // this.elements.forEach((multiplicity) => {
        //     len += multiplicity;
        // });
        // return len;
    }
    distinctElements() {
        return new Set(this.elements.keys());
    }
    has(element) {
        return this.elements.has(element);
    }
    add(element, multiplicity) {
        if (multiplicity > 0) {
            if (this.elements.has(element)) {
                var newMultiplicity = this.elements.get(element) + multiplicity;
                if (newMultiplicity == 0) {
                    this.elements.delete(element);
                } else {
                    this.elements.set(element, newMultiplicity);
                }
            } else {
                this.elements.set(element, multiplicity);
            }
        }
    }
    remove(element, multiplicity) {
        if (multiplicity > 0) {
            if (this.elements.has(element)) {
                const diff = this.elements.get(element) - multiplicity;
                if (diff > 0) {
                    this.elements.set(element, diff);
                } else {
                    this.elements.delete(element);
                }
            } else {
                this.elements.set(element, multiplicity);
            }
        }
    }
    combine(other) {
        if (other instanceof Multiset) {
            other.elements.forEach((ele, mul) => {
                if (this.elements.has(ele)) {
                    this.elements.set(ele, this.elements.get(ele) + mul);
                } else {
                    this.elements.set(ele, mul);
                }
            });
        }
    }
    count(element) {
        return !this.has(element) ? 0 : this.elements.get(element);
    }
    times(factor) {
        if (factor > 0) {
            this.elements.forEach((element, multiplicity) => {
                this.elements.set(element, multiplicity * factor);
            });
        }
    }
    isSubsetOf(other) {
        if (!other instanceof Multiset) {
            return false;
        }
        return Array.from(
            union(this.distinctElements(), other.distinctElements())
        ).every((element) => this.count(element) <= other.count(element));
    }
    isSupersetOf(other) {
        return other.isSubsetOf(this);
    }
    union(other) {
        if (other instanceof Multiset) {
            const u = union(this.distinctElements(), other.distinctElements());
            const m = new Map();
            u.forEach((i) => {
                m.set(i, Math.max(this.count(i), other.count(i)));
            });
            return new Multiset(m);
        }
    }
    intersection(other) {
        if (other instanceof Multiset) {
            const i = intersection(
                this.distinctElements(),
                other.distinctElements()
            );
            const m = new Map();
            i.forEach((i) => {
                m.set(i, Math.min(this.count(i), other.count(i)));
            });
            return new Multiset(m);
        }
    }
    difference(other) {
        if (other instanceof Multiset) {
            const i = union(this.distinctElements(), other.distinctElements());
            const m = new Map();
            i.forEach((i) => {
                m.set(i, Math.max(this.count(i) - other.count(i), 0));
            });
            return new Multiset(m);
        }
    }
    symmetricDifference(other) {
        if (other instanceof Multiset) {
            const i = union(this.distinctElements(), other.distinctElements());
            const m = new Map();
            i.forEach((e) => {
                m.set(e, Math.abs(this.count(e) - other.count(e)));
            });
            return new Multiset(m);
        }
    }
    forEach(func) {
        if (func instanceof Function) {
            this.elements.forEach((multiplicity, element) => {
                while (multiplicity > 0) {
                    func(element);
                    multiplicity--;
                }
            });
        }
    }
    all(boolfunc) {
        return (
            boolfunc instanceof Function &&
            this.isEmpty() &&
            Array.from(this.elements.keys()).every((e) => boolfunc(e))
        );
    }
    any(boolfunc) {
        return (
            boolfunc instanceof Function &&
            this.isEmpty() &&
            Array.from(this.elements.keys).some((e) => boolfunc(e))
        );
    }
}

class Hybridset {
    constructor(iterable) {
        if (iterable instanceof Map) {
            this.elements = new Map(
                Array.from(iterable.entries()).filter(
                    ([element, multiplicity]) => multiplicity != 0
                )
            );
        }
        if (iterable instanceof Hybridset) {
            this.elements = iterable.elements;
        }
        if (
            iterable instanceof Array &&
            iterable.length == 2 &&
            iterable.every((x) => x instanceof Multiset)
        ) {
            this.multisetGenerate(iterable[0], iterable[1]);
        }
    }
    copy() {
        return new Hybridset(this);
    }
    isEmpty() {
        return this.elements.size == 0;
    }
    clear() {
        this.elements.clear();
    }
    multisetGenerate(a, b) {
        if (this.isEmpty()) {
            this.clear();
        }
        if (a instanceof Set || a instanceof Array) {
            a.forEach((element) => {
                if (this.elements.has(element)) {
                    this.elements.set(element, this.elements.get(element) + 1);
                } else {
                    this.elements.set(element, 1);
                }
            });
        }
        if (a instanceof Multiset) {
            a.elements.combine(a);
        }
        if (b instanceof Set || b instanceof Array) {
            b.forEach((element) => {
                if (this.elements.has(element)) {
                    this.elements.set(element, this.elements.get(element) - 1);
                } else {
                    this.elements.set(element, -1);
                }
            });
        }
        if (b instanceof Multiset) {
            b.elements.combine(b);
        }
    }
    has(element) {
        return this.elements.has(element);
    }
    count(element) {
        return this.has(element) ? this.elements.get(element) : 0;
    }
    negativePart() {
        var newMap = new Map(this.elements);
        this.elements.forEach((multiplicity, element) => {
            if (multiplicity > 0) {
                newMap.delete(element);
            } else {
                newMap.set(element, -multiplicity);
            }
        });
        return new Multiset(newMap);
    }
    positivePart() {
        var newMap = new Map(this.elements);
        this.elements.forEach((multiplicity, element) => {
            if (multiplicity < 0) {
                newMap.delete(element);
            }
        });
        return new Multiset(newMap);
    }
    distinctElements() {
        return new Set(this.elements.keys());
    }
    toString() {
        let a = this.positivePart().toString();
        let b = this.negativePart().toString();
        return a.slice(0, -1) + "|" + b.slice(1);
    }
    cardinality() {
        var i = 0;
        this.elements.forEach((multiplicity, element) => {
            i += multiplicity;
        });
        return i;
    }
    weight() {
        var i = 0;
        this.elements.forEach((multiplicity, element) => {
            i += Math.abs(multiplicity);
        });
        return i;
    }
    add(element, multiplicity) {
        if (this.elements.has(element)) {
            var s = this.elements.get(element) + multiplicity;
            if (s == 0) {
                this.elements.delete(element);
            } else {
                this.elements.set(element, s);
            }
        } else {
            this.elements.set(element, multiplicity);
        }
    }
    remove(element, multiplicity) {
        if (this.elements.has(element)) {
            var s = this.elements.get(element) - multiplicity;
            if (s == 0) {
                this.elements.delete(element);
            } else {
                this.elements.set(element, s);
            }
        } else {
            this.elements.set(element, -multiplicity);
        }
    }
    combine(other) {
        if (other instanceof Multiset || other instanceof Hybridset) {
            const localMap = this.elements;
            other.elements.forEach((multiplicity, element) => {
                if (!localMap.has(element)) {
                    localMap.set(element, multiplicity);
                } else {
                    localMap.set(element, localMap.get(element) - multiplicity);
                }
            });
            return new Hybridset(localMap);
        }
    }
    times(factor) {
        let anotherHybridset = this.copy();
        anotherHybridset.elements.forEach((multiplicity, element) => {
            anotherHybridset.elements.set(element, multiplicity * factor);
        });
        return anotherHybridset;
    }
    part(other) {
        if (other instanceof Hybridset || other instanceof Multiset) {
            const localMap = this.elements;
            other.elements.forEach((multiplicity, element) => {
                if (!localMap.has(element)) {
                    localMap.set(element, -multiplicity);
                } else {
                    localMap.set(element, localMap.get(element) - multiplicity);
                }
            });
            return new Hybridset(localMap);
        }
    }
    isNewset() {
        const afte = Array.from(this.elements);
        return afte.every((m) => m[1] == -1) || afte.every((m) => m[1] == 1);
    }
    isProper() {
        return Array.from(this.elements).some((m) => m[1] < 0);
    }
    isSubsetOf(other) {
        return (
            (other instanceof Hybridset || other instanceof Multiset) &&
            Array.from(
                union(this.distinctElements(), other.distinctElements())
            ).every((m) => {
                const t = this.count(m),
                    o = other.count(m);
                return (t <= o && o < 0) || (0 <= t && t <= o);
            })
        );
    }
}

File: hybridset.py
from collections import defaultdict
from collections.abc import Mapping, Sized
from multiset import Multiset, FrozenMultiset
from itertools import chain, repeat, starmap

_sequence_types = (
    list,
    tuple,
    set,
    frozenset,
    range,
    str,
    Multiset,
    FrozenMultiset,
    Sized,
)
_iter_types = (type(iter([])), type((lambda: (yield))()))

__all__ = ["hybridset", "frozenhybridset"]


class hybridset(object):
    __slots__ = ("_elements", "_p", "_n")

    # Khởi tạo hybridset
    def __init__(self, iterable=None, another_iterable=None) -> None:
        self._elements = defaultdict(int)
        self._p = 0
        self._n = 0
        if isinstance(iterable, hybridset):
            self._elements = iterable._elements.copy()
            self._p = iterable._p
            self._n = iterable._n
        else:
            if iterable is not None:
                if another_iterable is not None:
                    if isinstance(iterable, _sequence_types) and isinstance(
                        another_iterable,
                        _sequence_types,
                    ):
                        for element in iterable:
                            self._elements[element] += 1
                        for element in another_iterable:
                            self._elements[element] -= 1

                        self._p = len(Multiset(iterable) - Multiset(another_iterable))
                        self._n = len(Multiset(another_iterable) - Multiset(iterable))
                    else:
                        raise TypeError(
                            "Two arguments to construct this hybridset must be a instance of one in the sequence types!"
                        )
                else:
                    if isinstance(iterable, (dict, Mapping)):
                        for element, multiplicity in iterable.items():
                            if type(multiplicity) == type(0):
                                pass
                            if multiplicity != 0:
                                self._elements[element] = multiplicity
                            if multiplicity > 0:
                                self._p += multiplicity
                            elif multiplicity < 0:
                                self._n += multiplicity
                    else:
                        for element in iterable:
                            self._elements[element] += 1
                            self._p += 1

    # Kiểm tra xem phần tử có tồn tại trong hybridst hay không
    def __contains__(self, element):
        return element in self._elements

    def __getitem__(self, element):
        return self._elements.get(element, 0)

    # In ra màn hình
    def __str__(self) -> str:
        return "{%s|%s}" % (
            ", ".join(map(str, self.positive_part().__iter__())),
            ", ".join(map(str, self.negative_part().__iter__())),
        )

    def __bool__(self):
        return self._p > 0 or self._n > 0

    # So sánh bằng với hai giá trị hybridset
    def __eq__(self, other):
        return (
            self._elements == other._elements if isinstance(other, hybridset) else False
        )

    # Sao chép
    def copy(self):
        return self.__class__(self)

    # Xóa toàn bộ phần tử
    def clear(self):
        self._elements.clear()
        self._p = 0
        self._n = 0

    __copy__ = copy

    # Phần dương và phần âm
    def positive_part(self):
        return Multiset(
            {
                element: self._elements[element]
                for element in self._elements
                if self._elements[element] > 0
            }
        )

    def negative_part(self):
        return Multiset(
            {
                element: -self._elements[element]
                for element in self._elements
                if self._elements[element] < 0
            }
        )

    # Các phần tử riêng biệt
    def distinct_elements(self):
        return set(self._elements.keys())

    # Lực lượng và trọng số
    def cardinality(self):
        return sum(self._elements.values())

    def weight(self):
        return sum(map(abs, self._elements.values()))

    # Nhân vô hướng với một số
    def times(self, factor):
        result = self.__copy__()
        if factor == 0:
            return self.__class__()
        elif factor < 0:
            result._p, result._n = result._n, result._p
        else:
            for element in result._elements:
                result._elements[element] *= factor
        result._n *= factor
        result._p *= factor
        return result

    def __mul__(self, factor):
        if not isinstance(factor, int):
            return NotImplemented
        return self.times(factor)

    def __imul__(self, factor):
        if not isinstance(factor, int):
            return NotImplemented
        return self.times(factor)

    def supportingsets(self):
        a = self.distinct_elements()
        return set(filter(lambda element: self._elements[element] > 0, a)), set(
            filter(lambda element: self._elements[element] < 0, a)
        )

    # Kiểm tra xem hybridset có phải là một new set hay không
    def isnewset(self):
        return all(self._elements[element] == 1 for element in self._elements) or all(
            self._elements[element] == -1 for element in self._elements
        )

    # Kiểm tra xem phần âm của hybridset có rỗng hay không
    def isproper(self):
        return any(self._elements[element] < 0 for element in self._elements)

    def __iter__(self):
        return chain.from_iterable(starmap(repeat, self._elements.items()))

    def isdisjoint(self, other):
        return all(element not in other for element in self._elements.keys())

    # Thêm và gỡ một phần tử ra khỏi hybridset
    def add(self, element, multiplicity=1):
        if multiplicity < 0:
            return
        a = self._elements[element]
        if a < 0:
            b = multiplicity - a
            if b > 0:
                self._p += b
                self._n -= a
            else:
                self._n -= multiplicity
        else:
            self._p += multiplicity
        self._elements[element] += multiplicity
        return

    def remove(self, element, multiplicity=1):
        if multiplicity < 0:
            return
        a = self._elements[element]
        if a > 0:
            b = multiplicity - a
            if b > 0:
                self._n += b
                self._p -= a
            else:
                self._p -= multiplicity
        else:
            self._n += multiplicity
        self._elements[element] -= multiplicity

    # Kết hợp hai hybridset
    def combine(self, other):
        if not isinstance(other, hybridset):
            return NotImplemented
        return hybridset(
            {
                element: self._elements[element] + other._elements[element]
                for element in self.distinct_elements() | other.distinct_elements()
                if self._elements[element] + other._elements[element] != 0
            }
        )

    def __add__(self, other):
        return self.combine(other)

    def __iadd__(self, other):
        return self.combine(other)

    def __neg__(self):
        return self * -1

    def __sub__(self, other):
        return self.combine(other * -1)

    def __isub__(self, other):
        return self.combine(other * -1)

    # Kiểm tra xem nó có phải là tập hợp lai con của một tập hợp lai cho trước hay không
    def issubset(self, other):
        if not isinstance(other, hybridset):
            return NotImplemented
        return all(
            self.partial_ordering(multiplicity, other._elements[element])
            or self.partial_ordering(
                other._elements[element] - multiplicity, other._elements[element]
            )
            for element, multiplicity in self._elements.items()
        )

    # Kiểm tra xem nó có "chứa" một tập hợp lai cho trước hay không
    def issuperset(self, other):
        return (
            NotImplemented
            if not isinstance(other, hybridset)
            else bool(other.issubset(self))
        )

    # Kiểm tra xem nó có phải là tập hợp lai con "tự nhiên" của một tập hợp lai cho trước hay không
    def isnaturalsubset(self, other):
        if not isinstance(other, hybridset):
            return NotImplemented
        a = other.supportingsets()
        return (
            self.distinct_elements().issubset(other.distinct_elements())
            and all(
                0 <= self._elements[element] <= other._elements[element]
                for element in a[0]
            )
            and all(self._elements[element] >= 0 for element in a[1])
        )

    # Phần bù của một tập hợp cho trước trên một tập hợp kia
    def complement(self, other):
        return (
            NotImplemented
            if not isinstance(other, hybridset)
            else other - self if self.isnaturalsubset(other) else None
        )

    @staticmethod
    def partial_ordering(a, b):
        return a <= b < 0 or 0 <= a <= b


class frozenhybridset(hybridset):
    __slots__ = ()

    def hash(self):
        return hash(frozenset(self._elements.items()))
